core/src/main/kotlin/com/tschuchort/compiletesting/KotlinJsCompilation.kt

```diff
@@ -0,0 +1,139 @@
+   * generates both IR and pre-IR versions of library.
+   */
+  var irProduceKlibDir: Boolean = false
+
+  /** Generate packed klib into file specified by -output. Disables pre-IR backend */
+  var irProduceKlibFile: Boolean = false
+
+  /** Generates JS file using IR backend. Also disables pre-IR backend */
+  var irProduceJs: Boolean = false
+
+  /** Perform experimental dead code elimination */
+  var irDce: Boolean = false
+
+  /** Perform a more experimental faster dead code elimination */
+  var irDceDriven: Boolean = false
+
+  /** Print declarations' reachability info to stdout during performing DCE */
+  var irDcePrintReachabilityInfo: Boolean = false
+
+  /** Disables pre-IR backend */
+  var irOnly: Boolean = false
+
+  /** Specify a compilation module name for IR backend */
+  var irModuleName: String? = null
+
+  /**
+   * Path to the kotlin-stdlib-js.jar
+   * If none is given, it will be searched for in the host
+   * process' classpaths
+   */
+  var kotlinStdLibJsJar: File? by default {
+    findInHostClasspath(hostClasspaths, "kotlin-stdlib-js.jar",
+      kotlinDependencyRegex("kotlin-stdlib-js"))
+  }
+
+  // *.class files, Jars and resources (non-temporary) that are created by the
+  // compilation will land here
+  val outputDir get() = workingDir.resolve("output")
+
+  /** Result of the compilation */
+  inner class Result(
+    /** The exit code of the compilation */
+    val exitCode: KotlinCompilation.ExitCode,
+    /** Messages that were printed by the compilation */
+    val messages: String
+  ) {
+    /** The directory where only the final output class and resources files will be */
+    val outputDirectory: File get() = outputDir
+
+    /**
+     * Compiled class and resource files that are the final result of the compilation.
+     */
+    val compiledClassAndResourceFiles: List<File> = outputDirectory.listFilesRecursively()
+  }
+
+
+  // setup common arguments for the two kotlinc calls
+  private fun commonK2JSArgs() = commonArguments(K2JSCompilerArguments()) { args ->
+    // the compiler should never look for stdlib or reflect in the
+    // kotlinHome directory (which is null anyway). We will put them
+    // in the classpath manually if they're needed
+    args.noStdlib = true
+
+    args.moduleKind = "commonjs"
+    args.outputFile = File(outputDir, outputFileName).absolutePath
+    args.sourceMapBaseDirs = commonClasspaths().joinToString(separator = File.pathSeparator)
+    args.libraries = listOfNotNull(kotlinStdLibJsJar).joinToString(separator = ":")
+
+    args.irProduceKlibDir = irProduceKlibDir
+    args.irProduceKlibFile = irProduceKlibFile
+    args.irProduceJs = irProduceJs
+    args.irDce = irDce
+    args.irDceDriven = irDceDriven
+    args.irDcePrintReachabilityInfo = irDcePrintReachabilityInfo
+    args.irOnly = irOnly
+    args.irModuleName = irModuleName
+  }
+
+  /** Runs the compilation task */
+  fun compile(): Result {
```

Comment:
There is a `Result` nested class for both KotlinJsCompilcation and KotlinCompilation. So the return type is different for each which is why it's a little hard to share. Ideally I would have pulled the Result class up to a top level class so it could be shared but that would break compatibility. I wasn't sure the best way to handle Result and ExitCode since they are both nested classes. Should I duplicate them both?

Reply:
In that case it's fine the way you did it.